smolyay/grid.py

```diff
@@ -0,0 +1,458 @@
+import sys
+sys.path.insert(1, '/home/che_h2/mzf0069/Documents/code/smolyak-surrogate-modeling')
+import abc
+import numpy
+import itertools
+from smolyay.basis import (BasisFunction, ChebyshevFirstKind,
+                   BasisFunctionSet, NestedBasisFunctionSet)
+
+
+class IndexGridGenerator(abc.ABC):
+    """Grid points for interpolation.
+
+    Depending on the approach and dimensionality
+    , a set of grid points, and their corresponding
+    basis functions are generated. These grid points
+    and functions can then be then used for approximation
+    of complex systems.
+    """
+
+    def __init__(self):
```

Comment:
I think you should add a `basis_set` argument here, save it to `self._basis_set`, then expose it as the `basis_set` property. All classes are likely to use it.